zcash_client_sqlite/src/testing.rs

```diff
@@ -25,19 +46,452 @@ use zcash_primitives::{
+use zcash_client_backend::data_api::wallet::shield_transparent_funds;
+#[cfg(feature = "transparent-inputs")]
+use zcash_primitives::{
+    legacy, legacy::keys::IncomingViewingKey, transaction::components::amount::NonNegativeAmount,
+};
 
-use crate::{wallet::init::init_accounts_table, AccountId, WalletDb};
+use crate::{
+    chain::init::init_cache_database,
+    error::SqliteClientError,
+    wallet::{
+        commitment_tree,
+        init::{init_accounts_table, init_wallet_db},
+        sapling::tests::test_prover,
+    },
+    AccountId, ReceivedNoteId, WalletDb,
+};
 
 use super::BlockDb;
 
 #[cfg(feature = "unstable")]
-use super::chain::BlockMeta;
+use crate::{
+    chain::{init::init_blockmeta_db, BlockMeta},
+    FsBlockDb,
+};
+/// A builder for a `zcash_client_sqlite` test.
+pub(crate) struct TestBuilder<Cache> {
+    cache: Cache,
+    seed: Option<SecretVec<u8>>,
+    with_test_account: bool,
+}
+
+impl TestBuilder<()> {
+    /// Constructs a new test.
+    pub(crate) fn new() -> Self {
+        TestBuilder {
+            cache: (),
+            seed: None,
+            with_test_account: false,
+        }
+    }
+
+    /// Adds a [`BlockDb`] cache to the test.
+    pub(crate) fn with_block_cache(self) -> TestBuilder<BlockCache> {
+        TestBuilder {
+            cache: BlockCache::new(),
+            seed: self.seed,
+            with_test_account: self.with_test_account,
+        }
+    }
+
+    /// Adds a [`FsBlockDb`] cache to the test.
+    #[cfg(feature = "unstable")]
+    pub(crate) fn with_fs_block_cache(self) -> TestBuilder<FsBlockCache> {
+        TestBuilder {
+            cache: FsBlockCache::new(),
+            seed: self.seed,
+            with_test_account: self.with_test_account,
+        }
+    }
+}
+
+impl<Cache> TestBuilder<Cache> {
+    /// Gives the test knowledge of the wallet seed for initialization.
+    pub(crate) fn with_seed(mut self, seed: SecretVec<u8>) -> Self {
+        // TODO remove
+        self.seed = Some(seed);
+        self
+    }
+
+    pub(crate) fn with_test_account(mut self) -> Self {
+        self.with_test_account = true;
+        self
+    }
+
+    /// Builds the runner for this test.
+    pub(crate) fn build(self) -> TestRunner<Cache> {
+        let params = network();
+
+        let data_file = NamedTempFile::new().unwrap();
+        let mut db_data = WalletDb::for_path(data_file.path(), params).unwrap();
+        init_wallet_db(&mut db_data, self.seed).unwrap();
+
+        let test_account = if self.with_test_account {
+            // Add an account to the wallet
+            Some(init_test_accounts_table_ufvk(&mut db_data))
+        } else {
+            None
+        };
+
+        TestRunner {
+            params,
+            cache: self.cache,
+            latest_cached_block: None,
+            _data_file: data_file,
+            db_data,
+            test_account,
+        }
+    }
+}
+
+/// A `zcash_client_sqlite` test runner.
+pub(crate) struct TestRunner<Cache> {
```

Comment:
This seems somewhat misnamed, because it doesn't actually run a test. Perhaps `TestContext` instead?

Reply:
I initially thought it might end up being more of a test runner / driver (with the wallet available through a closure similar to what you did in #915), but it ended up being a much simpler change to have it be a "test oracle". I'd weakly prefer `TestOracle` over `TestContext` myself, as it's not a passive context but an active participant in the test.

Reply:
Going with `TestState`.